File: mpp/Lab6/mpp-proiect-java/javaPb9/src/main/java/ro/mpp2024/repository/ProbaRepository.java
package ro.mpp2024.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.model.Arbitru;
import ro.mpp2024.model.Categorie;
import ro.mpp2024.model.Proba;
import ro.mpp2024.repository.interfaces.IProbaRepository;
import ro.mpp2024.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class ProbaRepository implements IProbaRepository {

    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();

    public ProbaRepository(JdbcUtils dbUtils) {
        logger.info("Initializing ProbaRepository with properties: {} ");
        this.dbUtils = dbUtils;
    }

    @Override
    public Proba createEntityFromResultSet(ResultSet resultSet) throws SQLException {
        logger.traceEntry("createEntityFromResultSet");
        String id = resultSet.getString("id");
        String nume = resultSet.getString("nume");
        Categorie categorie = Categorie.valueOf(resultSet.getString("categorie"));
        Proba proba = new Proba(id,nume,categorie);
        proba.setArbitru(resultSet.getLong("arbitru_probei"));
        logger.traceExit(proba);
        return proba;
    }

    @Override
    public Iterable<Proba> findAll() {
        logger.traceEntry("findAll");
        Connection con = dbUtils.getConnection();
        List<Proba> probe = new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select * from Proba")) {
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Proba proba = createEntityFromResultSet(result);
                    probe.add(proba);
                }
            }
        }catch (SQLException e){
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit(probe);
        return probe;
    }

    @Override
    public Proba findOne(String s) {
        logger.traceEntry("findOne");
        Connection con = dbUtils.getConnection();
        Proba proba = null;
        try(PreparedStatement preStmt=con.prepareStatement("select * from Proba where id=?")){
            preStmt.setString(1,s);
            try(ResultSet result=preStmt.executeQuery()){
                if(result.next()){
                    proba = createEntityFromResultSet(result);
                }
            }
        }catch (SQLException e){
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit(proba);
        return proba;
    }

    @Override
    public void save(Proba entity) {
        logger.traceEntry("save");
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into Proba (id,nume,arbitru_probei,categorie) values (?,?,?,?)")){
            Arbitru arbitru= getArbitru(entity.getId());
            preStmt.setString(1,entity.getId());
            preStmt.setString(2,entity.getNume());
            preStmt.setLong( 3, arbitru.getId());
            preStmt.setString(4,entity.getCategorie().toString());
            int result = preStmt.executeUpdate();
        }catch (SQLException e){
            logger.error(e);
            System.out.println("Error DB " + e);
        }
    }

    @Override
    public Proba delete(String s) {
        //  TODO: implement method when needed
        return null;
    }

    @Override
    public Proba update(String s, Proba new_entity) {
        //TODO: implement method when neeeded
        return null;
    }

    @Override
    public Arbitru getArbitru(String id) {
        logger.traceEntry("getArbitru");
        Arbitru arbitru = null;
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Arbitru where proba_asociata=?")) {
            preStmt.setString(1, id);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    arbitru = new Arbitru(result.getLong("id"), result.getString("nume"),
                            result.getString("username"), result.getString("parola")
                            , id);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit(arbitru);
        return arbitru;
    }
}

File: mpp/Lab11/csharpPb9/persistene/repository/ProbaRepository.cs
using csharpPb9.utils;
using log4net;
using System.Data;

public class ProbaRepository : IProbaRepository
{
    private static readonly ILog log = LogManager.GetLogger("");

    IDictionary<string, string> props;


    public ProbaRepository(IDictionary<string, string> props)
    {
        log.Info("Creating ProbaRepository with properties: " + props);
        this.props = props;
    }
    public Proba fromSettoEntity(IDataReader read)
    {
        string id = read.GetString(0);
        string nume = read.GetString(1);
        long id_arbitru = read.GetInt64(2);
        Categorie categorie = (Categorie)Enum.Parse(typeof(Categorie), read.GetString(3));
        Proba proba = new Proba(id, nume, categorie);
        proba.Id_arbitru = id_arbitru;
        return proba;
    }
    public IEnumerable<Proba> FindAll()
    {
        log.Info("Finding all probe");
        var con = DBUtils.getConnection(props);
        IList<Proba> probe = new List<Proba>();
        try
        {
            con.Open();
            using (IDbCommand conn = con.CreateCommand())
            {
                conn.CommandText = "SELECT * FROM Proba";
                using (IDataReader reader = conn.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        Proba proba = fromSettoEntity(reader);
                        probe.Add(proba);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }
        log.InfoFormat("Exiting FindAll with value {0}", probe);
        return probe;
    }

    public Arbitru FindArbitru(string id)
    {
        log.Info($"Find arbitru");
        var con = DBUtils.getConnection(props);
        try
        {
            con.Open();
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "SELECT * FROM Arbitru WHERE proba_asociata=@id";
                var param = comm.CreateParameter();
                param.ParameterName = "@id";
                param.Value = id;
                comm.Parameters.Add(param);
                using (IDataReader reader = comm.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        log.Info("Found Arbitru");
                        return new Arbitru(reader.GetInt64(0), reader.GetString(1), reader.GetString(2), reader.GetString(3), reader.GetString(4));
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }

        log.Info("No Arbitru found");
        return null;
    }

    public Proba FindOne(string id)
    {
       log.Info($"Finding proba with id {id}");
        var con = DBUtils.getConnection(props);
        Proba proba = null;
        try
        {
            con.Open();
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "SELECT * FROM Proba WHERE id=@id";
                var param = comm.CreateParameter();
                param.ParameterName = "@id";
                param.Value = id;
                comm.Parameters.Add(param);
                using (IDataReader reader = comm.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        proba = fromSettoEntity(reader);
                    }
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }

        log.InfoFormat("Exiting FindOne with value {0}", proba);
        return proba;
    }

    public void Save(Proba entity)
    {
        log.Info("Saving proba");
        var con = DBUtils.getConnection(props);
        try
        {
            con.Open();
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "INSERT INTO Proba(id, nume, arbitru_probei, categorie) VALUES (@id, @nume, @id_arbitru, @categorie)";
                #region vars
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = entity.Id;
                comm.Parameters.Add(paramId);

                var paramNume = comm.CreateParameter();
                paramNume.ParameterName = "@nume";
                paramNume.Value = entity.Nume;
                comm.Parameters.Add(paramNume);

                var paramIdArbitru = comm.CreateParameter();
                paramIdArbitru.ParameterName = "@id_arbitru";
                paramIdArbitru.Value = entity.Id_arbitru;
                comm.Parameters.Add(paramIdArbitru);

                var paramCategorie = comm.CreateParameter();
                paramCategorie.ParameterName = "@categorie";
                paramCategorie.Value = entity.Categorie.ToString();
                comm.Parameters.Add(paramCategorie);

                #endregion
                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    log.Error("No proba saved!");
                }
                else
                {
                    log.Info("Proba saved");
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }
        log.Info("Exiting Save");
    }

    public void SetArbitruForProba(string id, long arbitru)
    {
        log.Info("Setting arbitru for proba");
        var con = DBUtils.getConnection(props);
        try
        {
            con.Open();
            using (IDbCommand comm = con.CreateCommand())
            {
                comm.CommandText = "UPDATE Proba SET arbitru_probei=@arbitru WHERE id=@id";
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);
                var paramArbitru = comm.CreateParameter();
                paramArbitru.ParameterName = "@arbitru";
                paramArbitru.Value = arbitru;
                comm.Parameters.Add(paramArbitru);
                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    log.Error("No proba updated!");
                }
                else
                {
                    log.Info("Proba updated");
                }
            }
        }
        catch (Exception e)
        {
            log.Error(e);
            Console.WriteLine(e.Message);
        }
        finally
        {
            log.Info("Closing connection");
            con.Close();
        }
    }

    #region To implement
    public Proba Delete(string id)
    {
        throw new NotImplementedException();
    }

    public Proba Update(string id, Proba new_entity)
    {
        throw new NotImplementedException();
    }

    #endregion
}